In [1]:
!python -m pip install -U python-woc pandas matplotlib
# please clear the output of this cell in your notebook before checking it in

Defaulting to user installation because normal site-packages is not writeable


In [2]:
import sys
print(sys.executable)

/usr/bin/python3.11


In [3]:
import sys
!{sys.executable} -m pip install -U python-woc

Defaulting to user installation because normal site-packages is not writeable


In [4]:
from woc.remote import WocMapsRemote
from tqdm import tqdm
import pandas as pd
# creates the client
woc = WocMapsRemote( base_url="https://worldofcode.org/api/")
# if you got an API key
# woc = WocMapsRemote( base_url="https://worldofcode.org/api/", api_key="woc-XXXXXX-YYYYYY" )

# Now for each of the ten projects you were assigned get commits, e.g.
projects = ['delsuc/spike',
            'gyyang/neurogym',
            'PapenfussLab/gridss',
            'jbytecode/LinRegOutliers',
            'nasa/CCDD',
            'arm61/pylj',
            'mothur/mothur',
            'stephenslab/mashr',
            'dswah/pyGAM',
            'lsst/daf_butler'
           ]
list_df_commits = []
for prj in projects:
  # Convert GitHub repository names to WoC V2412 project names
  prj = prj.lower().replace('/','_',2)
  commits = woc.get_values('p2c', prj)
  df = pd.DataFrame(commits, columns=["sha1"])
  df['project'] = prj
  list_df_commits.append(df)
df = pd.concat(list_df_commits)
#perhaps save the list (if no errors) so you do not need to retrieve them again
df.to_csv('df_commits.csv')
df.head(1)

,sha1,project
0,004a56b528d6d1de376454a8cb0fa6701dc9ea71,delsuc_spike


In [5]:
# If the number of commits its not very large, you can try to get them all at the same time,
# The max batch size is 10
import time
# let us first split df['sha1'] in chunks
chunks = [df['sha1'][x:x+10] for x in range(0, len(df), 10)]
commit_data = []

for chunk in tqdm(chunks): # iterate over the commits
  # res, err = woc.show_content_many('commit',chunk.to_list())

  # commit.tch returns the same data but faster
  res, err = woc.get_values_many('commit.tch',chunk.to_list())
  res = {k: v[0] for k, v in res.items()}  # this conversion is necessary because of the internal implementation

  # to walk around the rate limit
  time.sleep(1)

  if err: # check for errors
    print('Got Errors', err)

  for commit_sha, commit in res.items():
    # flatten commit objects
    commit_data.append({
          'commit': commit_sha,
          'tree': commit[0],
          'parent': list(commit[1]),
          'author': commit[2][0],
          'author_time': int(commit[2][1]),
          'author_tz': commit[2][2],
          'committer': commit[3][0],
          'committer_time': int(commit[3][1]),
          'committer_tz': commit[3][2],
          'message': commit[4],
    })

df_commit_data = pd.DataFrame(commit_data)
df_commit_data = df_commit_data.merge(df, left_on='commit', right_on='sha1')
df_commit_data.to_csv('df_commit_data.csv', index=False)
df_commit_data.head(2)

 24%|██████████████████████▏                                                                     | 578/2395 [09:59<31:21,  1.04s/it]

Got Errors {'afde90a487c9f14697a93375d7e1fd964f0f01d8': 'Key afde90a487c9f14697a93375d7e1fd964f0f01d8 not found in /da5_fast/All.sha1c/commit_47.tch'}


 36%|█████████████████████████████████▎                                                          | 866/2395 [14:58<26:30,  1.04s/it]

Got Errors {'0db8ed378cf6560fd58a4c76f2b4582446827bea': 'Key 0db8ed378cf6560fd58a4c76f2b4582446827bea not found in /da5_fast/All.sha1c/commit_13.tch'}


 37%|█████████████████████████████████▋                                                          | 876/2395 [15:09<26:21,  1.04s/it]

Got Errors {'1348bc3893a79979292048f67b5ee3faa484f717': 'Key 1348bc3893a79979292048f67b5ee3faa484f717 not found in /da5_fast/All.sha1c/commit_19.tch'}


 37%|██████████████████████████████████▎                                                         | 893/2395 [15:26<25:52,  1.03s/it]

Got Errors {'1b900b8e71aae39564c0d48a2c187772d2e5dfb7': 'Key 1b900b8e71aae39564c0d48a2c187772d2e5dfb7 not found in /da5_fast/All.sha1c/commit_27.tch'}


 44%|███████████████████████████████████████▋                                                   | 1044/2395 [18:02<23:22,  1.04s/it]

Got Errors {'70e5f53ceb2082828664e80074700a786044d21a': 'Key 70e5f53ceb2082828664e80074700a786044d21a not found in /da5_fast/All.sha1c/commit_112.tch'}


 45%|████████████████████████████████████████▋                                                  | 1070/2395 [18:29<22:57,  1.04s/it]

Got Errors {'7e23ab9dedbf83a134675376c5b162144a69eea1': 'Key 7e23ab9dedbf83a134675376c5b162144a69eea1 not found in /da5_fast/All.sha1c/commit_126.tch'}


 55%|██████████████████████████████████████████████████▏                                        | 1321/2395 [22:48<18:34,  1.04s/it]

Got Errors {'2ea2c75f31cee58fb0dd84ca51d3d8b27669a573': 'Key 2ea2c75f31cee58fb0dd84ca51d3d8b27669a573 not found in /da5_fast/All.sha1c/commit_46.tch'}


 55%|██████████████████████████████████████████████████▏                                        | 1322/2395 [22:49<18:36,  1.04s/it]

Got Errors {'321af663d3ca2f1df7fad9d5579f1f6df21f9d4c': 'Key 321af663d3ca2f1df7fad9d5579f1f6df21f9d4c not found in /da5_fast/All.sha1c/commit_50.tch'}


 56%|███████████████████████████████████████████████████▍                                       | 1353/2395 [23:21<18:05,  1.04s/it]

Got Errors {'8c2c1d6089747cee6e47d168cb543f5111318bef': 'Key 8c2c1d6089747cee6e47d168cb543f5111318bef not found in /da5_fast/All.sha1c/commit_12.tch'}


 58%|████████████████████████████████████████████████████▎                                      | 1378/2395 [23:47<17:37,  1.04s/it]

Got Errors {'ccd3c241f1a8452f0cad794716a2e2435410c170': 'Key ccd3c241f1a8452f0cad794716a2e2435410c170 not found in /da5_fast/All.sha1c/commit_76.tch'}


 58%|████████████████████████████████████████████████████▌                                      | 1382/2395 [23:51<17:30,  1.04s/it]

Got Errors {'d5ccaa0b0a2da581d9283bf106b608d3c750540e': 'Key d5ccaa0b0a2da581d9283bf106b608d3c750540e not found in /da5_fast/All.sha1c/commit_85.tch'}


 61%|███████████████████████████████████████████████████████▋                                   | 1466/2395 [25:19<16:07,  1.04s/it]

Got Errors {'9c0d4a1ccbcb7fadab6f02bc0d846bc8bd0c9e09': 'Key 9c0d4a1ccbcb7fadab6f02bc0d846bc8bd0c9e09 not found in /da5_fast/All.sha1c/commit_28.tch'}


 62%|████████████████████████████████████████████████████████▍                                  | 1484/2395 [25:37<15:46,  1.04s/it]

Got Errors {'c99442e92d8528a268b0b5026afd5b548835c05e': 'Key c99442e92d8528a268b0b5026afd5b548835c05e not found in /da5_fast/All.sha1c/commit_73.tch'}


 64%|██████████████████████████████████████████████████████████                                 | 1527/2395 [26:22<15:04,  1.04s/it]

Got Errors {'06a3e51260d5f59c33be1154199df76ccd46c806': 'Key 06a3e51260d5f59c33be1154199df76ccd46c806 not found in /da5_fast/All.sha1c/commit_6.tch'}


 64%|██████████████████████████████████████████████████████████▎                                | 1535/2395 [26:30<14:57,  1.04s/it]

Got Errors {'08cefb5640895779113e4ed1b7ad084374270874': 'Key 08cefb5640895779113e4ed1b7ad084374270874 not found in /da5_fast/All.sha1c/commit_8.tch'}


 65%|███████████████████████████████████████████████████████████                                | 1553/2395 [26:49<14:39,  1.04s/it]

Got Errors {'0dabc17a6a542dc9e98860b436743f370d456e7e': 'Key 0dabc17a6a542dc9e98860b436743f370d456e7e not found in /da5_fast/All.sha1c/commit_13.tch'}


 66%|███████████████████████████████████████████████████████████▋                               | 1571/2395 [27:08<14:19,  1.04s/it]

Got Errors {'1245a0ea6222083b797d885a897e8ef4958a8ea8': 'Key 1245a0ea6222083b797d885a897e8ef4958a8ea8 not found in /da5_fast/All.sha1c/commit_18.tch'}


 66%|████████████████████████████████████████████████████████████                               | 1582/2395 [27:19<14:04,  1.04s/it]

Got Errors {'155d718000639d8400a1b14d380e15fedaf65c74': 'Key 155d718000639d8400a1b14d380e15fedaf65c74 not found in /da5_fast/All.sha1c/commit_21.tch'}


 67%|█████████████████████████████████████████████████████████████▍                             | 1616/2395 [27:55<13:31,  1.04s/it]

Got Errors {'20192c5453d7941140f402ec637a17cfc230fbfe': 'Key 20192c5453d7941140f402ec637a17cfc230fbfe not found in /da5_fast/All.sha1c/commit_32.tch'}


 68%|█████████████████████████████████████████████████████████████▍                             | 1618/2395 [27:57<13:28,  1.04s/it]

Got Errors {'205d326b7128a2dc4cd416d276005e54a1c08a9f': 'Key 205d326b7128a2dc4cd416d276005e54a1c08a9f not found in /da5_fast/All.sha1c/commit_32.tch'}


 68%|█████████████████████████████████████████████████████████████▊                             | 1628/2395 [28:07<13:22,  1.05s/it]

Got Errors {'233a983462d89a1b7ff933af13a3a9bde1bd39f8': 'Key 233a983462d89a1b7ff933af13a3a9bde1bd39f8 not found in /da5_fast/All.sha1c/commit_35.tch'}


 68%|██████████████████████████████████████████████████████████████▎                            | 1640/2395 [28:20<13:08,  1.04s/it]

Got Errors {'2682293a478c1c2eb4d7f6ef286d12dd88bab0f1': 'Key 2682293a478c1c2eb4d7f6ef286d12dd88bab0f1 not found in /da5_fast/All.sha1c/commit_38.tch'}


 70%|███████████████████████████████████████████████████████████████▍                           | 1671/2395 [28:52<12:33,  1.04s/it]

Got Errors {'2f7d201d5b7ce9098ca04217488c2131dd5871fd': 'Key 2f7d201d5b7ce9098ca04217488c2131dd5871fd not found in /da5_fast/All.sha1c/commit_47.tch'}


 71%|████████████████████████████████████████████████████████████████▋                          | 1703/2395 [29:26<12:06,  1.05s/it]

Got Errors {'38394130cfb7e9b9e74f387ae820382b16197c77': 'Key 38394130cfb7e9b9e74f387ae820382b16197c77 not found in /da5_fast/All.sha1c/commit_56.tch'}


 71%|████████████████████████████████████████████████████████████████▋                          | 1704/2395 [29:27<12:02,  1.05s/it]

Got Errors {'387f3e0b5b1be08d3e182a37434b31dddde3f624': 'Key 387f3e0b5b1be08d3e182a37434b31dddde3f624 not found in /da5_fast/All.sha1c/commit_56.tch'}


 71%|████████████████████████████████████████████████████████████████▉                          | 1709/2395 [29:32<11:54,  1.04s/it]

Got Errors {'39c4f27e409606749c2831ec596ce7471a06bac3': 'Key 39c4f27e409606749c2831ec596ce7471a06bac3 not found in /da5_fast/All.sha1c/commit_57.tch'}


 71%|█████████████████████████████████████████████████████████████████                          | 1712/2395 [29:35<11:52,  1.04s/it]

Got Errors {'3a5a2a1834b37339e39b747b9ffc6977104b2d52': 'Key 3a5a2a1834b37339e39b747b9ffc6977104b2d52 not found in /da5_fast/All.sha1c/commit_58.tch'}


 72%|█████████████████████████████████████████████████████████████████▎                         | 1720/2395 [29:43<11:42,  1.04s/it]

Got Errors {'3c7dd09c01c6ec33cebba82501c22d4c8d2e6be4': 'Key 3c7dd09c01c6ec33cebba82501c22d4c8d2e6be4 not found in /da5_fast/All.sha1c/commit_60.tch'}


 72%|█████████████████████████████████████████████████████████████████▋                         | 1730/2395 [29:54<11:32,  1.04s/it]

Got Errors {'3f27e6073039078d614debec7ac1f3b6daa6bb02': 'Key 3f27e6073039078d614debec7ac1f3b6daa6bb02 not found in /da5_fast/All.sha1c/commit_63.tch'}


 72%|█████████████████████████████████████████████████████████████████▊                         | 1731/2395 [29:55<11:29,  1.04s/it]

Got Errors {'3f889c43e2cc0fbfe8d3de1d7e1c3c919fe27d0b': 'Key 3f889c43e2cc0fbfe8d3de1d7e1c3c919fe27d0b not found in /da5_fast/All.sha1c/commit_63.tch'}


 73%|██████████████████████████████████████████████████████████████████▏                        | 1741/2395 [30:05<11:20,  1.04s/it]

Got Errors {'425181977fa3b5234bb1d7ad7a43aa21aaefe0dd': 'Key 425181977fa3b5234bb1d7ad7a43aa21aaefe0dd not found in /da5_fast/All.sha1c/commit_66.tch'}


 73%|██████████████████████████████████████████████████████████████████▋                        | 1755/2395 [30:20<11:03,  1.04s/it]

Got Errors {'45f019c3c4073df5c3a6fbb1f8635578f7174f12': 'Key 45f019c3c4073df5c3a6fbb1f8635578f7174f12 not found in /da5_fast/All.sha1c/commit_69.tch'}


 73%|██████████████████████████████████████████████████████████████████▊                        | 1758/2395 [30:23<11:01,  1.04s/it]

Got Errors {'46ae4a316d6ba2b63efc107a9ccb7e6945423303': 'Key 46ae4a316d6ba2b63efc107a9ccb7e6945423303 not found in /da5_fast/All.sha1c/commit_70.tch'}


 74%|███████████████████████████████████████████████████████████████████▍                       | 1774/2395 [30:40<10:48,  1.04s/it]

Got Errors {'4b7ca00c3494a4622ecdcc2b4e046fa9426835c8': 'Key 4b7ca00c3494a4622ecdcc2b4e046fa9426835c8 not found in /da5_fast/All.sha1c/commit_75.tch'}


 74%|███████████████████████████████████████████████████████████████████▍                       | 1776/2395 [30:42<10:47,  1.05s/it]

Got Errors {'4c21b91327fa4821bcb04848100d82d32d4eb9ac': 'Key 4c21b91327fa4821bcb04848100d82d32d4eb9ac not found in /da5_fast/All.sha1c/commit_76.tch'}


 75%|████████████████████████████████████████████████████████████████████▍                      | 1801/2395 [31:08<10:21,  1.05s/it]

Got Errors {'538479c245f253e11e029ae5057e6a02899a631b': 'Key 538479c245f253e11e029ae5057e6a02899a631b not found in /da5_fast/All.sha1c/commit_83.tch'}


 75%|████████████████████████████████████████████████████████████████████▌                      | 1805/2395 [31:12<10:15,  1.04s/it]

Got Errors {'546e1345d71612a04f10119650588a9807da3ee6': 'Key 546e1345d71612a04f10119650588a9807da3ee6 not found in /da5_fast/All.sha1c/commit_84.tch'}


 75%|████████████████████████████████████████████████████████████████████▋                      | 1808/2395 [31:15<10:12,  1.04s/it]

Got Errors {'557792e1058e00d21173bb59c87eee1984fcf278': 'Key 557792e1058e00d21173bb59c87eee1984fcf278 not found in /da5_fast/All.sha1c/commit_85.tch'}


 76%|████████████████████████████████████████████████████████████████████▊                      | 1811/2395 [31:18<10:08,  1.04s/it]

Got Errors {'566bc069c91251065cb65deb0b2e7d000e6f8104': 'Key 566bc069c91251065cb65deb0b2e7d000e6f8104 not found in /da5_fast/All.sha1c/commit_86.tch'}


 76%|█████████████████████████████████████████████████████████████████████▌                     | 1831/2395 [31:39<09:48,  1.04s/it]

Got Errors {'5cc3e735235dd72917ea50fc40fe59aad335b7ab': 'Key 5cc3e735235dd72917ea50fc40fe59aad335b7ab not found in /da5_fast/All.sha1c/commit_92.tch'}


 77%|█████████████████████████████████████████████████████████████████████▉                     | 1840/2395 [31:48<09:39,  1.04s/it]

Got Errors {'5f71cd6d6303b06e9b10f86caba975249ad2428f': 'Key 5f71cd6d6303b06e9b10f86caba975249ad2428f not found in /da5_fast/All.sha1c/commit_95.tch'}


 77%|██████████████████████████████████████████████████████████████████████▏                    | 1847/2395 [31:56<09:29,  1.04s/it]

Got Errors {'61109e56ddeb9bbd028b54953af862887b6931cf': 'Key 61109e56ddeb9bbd028b54953af862887b6931cf not found in /da5_fast/All.sha1c/commit_97.tch', '61281467bd6248b8b2531da5ed7197d14662f6e1': 'Key 61281467bd6248b8b2531da5ed7197d14662f6e1 not found in /da5_fast/All.sha1c/commit_97.tch'}


 77%|██████████████████████████████████████████████████████████████████████▎                    | 1851/2395 [32:00<09:24,  1.04s/it]

Got Errors {'6295377a1424cf3440b5744aba9a88f3ed08200d': 'Key 6295377a1424cf3440b5744aba9a88f3ed08200d not found in /da5_fast/All.sha1c/commit_98.tch'}


 77%|██████████████████████████████████████████████████████████████████████▍                    | 1854/2395 [32:03<09:22,  1.04s/it]

Got Errors {'63057d226803978768d8945f3951f3269ac256ae': 'Key 63057d226803978768d8945f3951f3269ac256ae not found in /da5_fast/All.sha1c/commit_99.tch'}


 78%|██████████████████████████████████████████████████████████████████████▉                    | 1868/2395 [32:18<09:09,  1.04s/it]

Got Errors {'6645237af3731f5705de3e24274bbdb1f2821b66': 'Key 6645237af3731f5705de3e24274bbdb1f2821b66 not found in /da5_fast/All.sha1c/commit_102.tch'}


 78%|███████████████████████████████████████████████████████████████████████▏                   | 1875/2395 [32:25<09:03,  1.05s/it]

Got Errors {'6851adaead048f99ed8bc16fc326ea4f7e1fd9ff': 'Key 6851adaead048f99ed8bc16fc326ea4f7e1fd9ff not found in /da5_fast/All.sha1c/commit_104.tch'}


 79%|███████████████████████████████████████████████████████████████████████▋                   | 1887/2395 [32:37<08:49,  1.04s/it]

Got Errors {'6b7b5eecb500cb8f64cce054224b27ff91422a3a': 'Key 6b7b5eecb500cb8f64cce054224b27ff91422a3a not found in /da5_fast/All.sha1c/commit_107.tch'}


 80%|████████████████████████████████████████████████████████████████████████▌                  | 1910/2395 [33:01<08:28,  1.05s/it]

Got Errors {'7231cd11cf663ba7c7f858a6ac651530ce7ba5f5': 'Key 7231cd11cf663ba7c7f858a6ac651530ce7ba5f5 not found in /da5_fast/All.sha1c/commit_114.tch'}


 80%|████████████████████████████████████████████████████████████████████████▊                  | 1916/2395 [33:08<08:18,  1.04s/it]

Got Errors {'73e0e849d061cf6be0b13097ee76fc91e34b1802': 'Key 73e0e849d061cf6be0b13097ee76fc91e34b1802 not found in /da5_fast/All.sha1c/commit_115.tch'}


 80%|█████████████████████████████████████████████████████████████████████████                  | 1922/2395 [33:14<08:13,  1.04s/it]

Got Errors {'75cff5632d41a290e5b69eb3ef6f60204eff0087': 'Key 75cff5632d41a290e5b69eb3ef6f60204eff0087 not found in /da5_fast/All.sha1c/commit_117.tch'}


 81%|█████████████████████████████████████████████████████████████████████████▍                 | 1932/2395 [33:24<08:01,  1.04s/it]

Got Errors {'79029e154f0180b83075f92f849490a4ab2b4e3a': 'Key 79029e154f0180b83075f92f849490a4ab2b4e3a not found in /da5_fast/All.sha1c/commit_121.tch'}


 81%|█████████████████████████████████████████████████████████████████████████▊                 | 1943/2395 [33:36<07:51,  1.04s/it]

Got Errors {'7c297c0bcb59000d7adcb777f6482d2cdba7468b': 'Key 7c297c0bcb59000d7adcb777f6482d2cdba7468b not found in /da5_fast/All.sha1c/commit_124.tch'}


 82%|██████████████████████████████████████████████████████████████████████████▌                | 1964/2395 [33:58<07:26,  1.04s/it]

Got Errors {'821887ce78036ebba0a8c16943f8351ad0ea1ff3': 'Key 821887ce78036ebba0a8c16943f8351ad0ea1ff3 not found in /da5_fast/All.sha1c/commit_2.tch'}


 82%|███████████████████████████████████████████████████████████████████████████                | 1974/2395 [34:08<07:16,  1.04s/it]

Got Errors {'84ba4cb119651af44ae3aa115c6797f1c420f370': 'Key 84ba4cb119651af44ae3aa115c6797f1c420f370 not found in /da5_fast/All.sha1c/commit_4.tch'}


 82%|███████████████████████████████████████████████████████████████████████████                | 1975/2395 [34:09<07:15,  1.04s/it]

Got Errors {'8515214cbde458878f407db80ebff0a3777d5dd9': 'Key 8515214cbde458878f407db80ebff0a3777d5dd9 not found in /da5_fast/All.sha1c/commit_5.tch'}


 83%|███████████████████████████████████████████████████████████████████████████                | 1977/2395 [34:11<07:13,  1.04s/it]

Got Errors {'85a231b5fcac429a7fca47f1b4f54126523aa680': 'Key 85a231b5fcac429a7fca47f1b4f54126523aa680 not found in /da5_fast/All.sha1c/commit_5.tch'}


 83%|███████████████████████████████████████████████████████████████████████████▎               | 1983/2395 [34:17<07:08,  1.04s/it]

Got Errors {'8786d9f0dd30c60ed056aa1a5f78714ad6f5007f': 'Key 8786d9f0dd30c60ed056aa1a5f78714ad6f5007f not found in /da5_fast/All.sha1c/commit_7.tch'}


 83%|███████████████████████████████████████████████████████████████████████████▍               | 1985/2395 [34:19<07:04,  1.04s/it]

Got Errors {'87d7467a46b99da3f3710250f41917a2dbd0ab84': 'Key 87d7467a46b99da3f3710250f41917a2dbd0ab84 not found in /da5_fast/All.sha1c/commit_7.tch'}


 83%|███████████████████████████████████████████████████████████████████████████▍               | 1987/2395 [34:21<07:03,  1.04s/it]

Got Errors {'88d01c13954b51adfa83c13bd8a2e7e083981438': 'Key 88d01c13954b51adfa83c13bd8a2e7e083981438 not found in /da5_fast/All.sha1c/commit_8.tch'}


 84%|████████████████████████████████████████████████████████████████████████████▍              | 2012/2395 [34:47<06:37,  1.04s/it]

Got Errors {'907b86b5a30f90de372e1f277aec3abae84c7c6d': 'Key 907b86b5a30f90de372e1f277aec3abae84c7c6d not found in /da5_fast/All.sha1c/commit_16.tch', '908e5c5288cbf4f262a3324bf73b63d45bfc4dba': 'Key 908e5c5288cbf4f262a3324bf73b63d45bfc4dba not found in /da5_fast/All.sha1c/commit_16.tch'}


 85%|█████████████████████████████████████████████████████████████████████████████▏             | 2033/2395 [35:09<06:16,  1.04s/it]

Got Errors {'96ba999d0aed7e58cd3b9a7c89c85c46b649b2f5': 'Key 96ba999d0aed7e58cd3b9a7c89c85c46b649b2f5 not found in /da5_fast/All.sha1c/commit_22.tch'}


 85%|█████████████████████████████████████████████████████████████████████████████▍             | 2037/2395 [35:13<06:12,  1.04s/it]

Got Errors {'97e8798d32308cd0a81153c2d796e3701c8bd3e1': 'Key 97e8798d32308cd0a81153c2d796e3701c8bd3e1 not found in /da5_fast/All.sha1c/commit_23.tch'}


 86%|██████████████████████████████████████████████████████████████████████████████▌            | 2067/2395 [35:45<05:42,  1.04s/it]

Got Errors {'a1575ca21366aa541f4dd2ea4a7ea744b50c23f9': 'Key a1575ca21366aa541f4dd2ea4a7ea744b50c23f9 not found in /da5_fast/All.sha1c/commit_33.tch'}


 86%|██████████████████████████████████████████████████████████████████████████████▋            | 2070/2395 [35:48<05:38,  1.04s/it]

Got Errors {'a259b9b350839d93251425d1bdd7ae13b147293d': 'Key a259b9b350839d93251425d1bdd7ae13b147293d not found in /da5_fast/All.sha1c/commit_34.tch'}


 87%|██████████████████████████████████████████████████████████████████████████████▉            | 2079/2395 [35:57<05:28,  1.04s/it]

Got Errors {'a4ace9fee08ea2d1ea8f593b68565bd0f72f7c7f': 'Key a4ace9fee08ea2d1ea8f593b68565bd0f72f7c7f not found in /da5_fast/All.sha1c/commit_36.tch'}


 87%|███████████████████████████████████████████████████████████████████████████████▎           | 2086/2395 [36:04<05:21,  1.04s/it]

Got Errors {'a6b68af460c7de41cf935b8040c24b568e7f30b1': 'Key a6b68af460c7de41cf935b8040c24b568e7f30b1 not found in /da5_fast/All.sha1c/commit_38.tch'}


 87%|███████████████████████████████████████████████████████████████████████████████▎           | 2089/2395 [36:07<05:18,  1.04s/it]

Got Errors {'a79cb6ef93dde79a5704b553a340abc187298aec': 'Key a79cb6ef93dde79a5704b553a340abc187298aec not found in /da5_fast/All.sha1c/commit_39.tch'}


 87%|███████████████████████████████████████████████████████████████████████████████▍           | 2090/2395 [36:08<05:17,  1.04s/it]

Got Errors {'a7cf6e2e4f64248659e65519ae0e64d441e14bcc': 'Key a7cf6e2e4f64248659e65519ae0e64d441e14bcc not found in /da5_fast/All.sha1c/commit_39.tch'}


 89%|█████████████████████████████████████████████████████████████████████████████████▏         | 2137/2395 [36:57<04:28,  1.04s/it]

Got Errors {'b54f0003fe17c043c8d466e4ca7dd192f37187c3': 'Key b54f0003fe17c043c8d466e4ca7dd192f37187c3 not found in /da5_fast/All.sha1c/commit_53.tch'}


 90%|██████████████████████████████████████████████████████████████████████████████████▏        | 2163/2395 [37:24<04:01,  1.04s/it]

Got Errors {'bc5ae0764dfdd8caa68c523e188f4d31400e333b': 'Key bc5ae0764dfdd8caa68c523e188f4d31400e333b not found in /da5_fast/All.sha1c/commit_60.tch'}


 90%|██████████████████████████████████████████████████████████████████████████████████▎        | 2165/2395 [37:26<03:58,  1.04s/it]

Got Errors {'bcfad3d08a95d5efe8eb773f75c929d7f311dedf': 'Key bcfad3d08a95d5efe8eb773f75c929d7f311dedf not found in /da5_fast/All.sha1c/commit_60.tch'}


 91%|██████████████████████████████████████████████████████████████████████████████████▍        | 2171/2395 [37:33<03:52,  1.04s/it]

Got Errors {'be73e59282fc7ea1e0a12c03277c27aa1beb2b20': 'Key be73e59282fc7ea1e0a12c03277c27aa1beb2b20 not found in /da5_fast/All.sha1c/commit_62.tch'}


 91%|██████████████████████████████████████████████████████████████████████████████████▋        | 2176/2395 [37:38<03:47,  1.04s/it]

Got Errors {'bfbaefc08c34e35e2eabb65d06871cacbd8c8379': 'Key bfbaefc08c34e35e2eabb65d06871cacbd8c8379 not found in /da5_fast/All.sha1c/commit_63.tch'}


 91%|███████████████████████████████████████████████████████████████████████████████████▏       | 2188/2395 [37:50<03:34,  1.04s/it]

Got Errors {'c3048e3138427eaf70d809550cbac2bbbe790a59': 'Key c3048e3138427eaf70d809550cbac2bbbe790a59 not found in /da5_fast/All.sha1c/commit_67.tch'}


 92%|███████████████████████████████████████████████████████████████████████████████████▎       | 2193/2395 [37:55<03:29,  1.04s/it]

Got Errors {'c46186f82e0eb9b5240d04a06ddeb37c6dca3dad': 'Key c46186f82e0eb9b5240d04a06ddeb37c6dca3dad not found in /da5_fast/All.sha1c/commit_68.tch'}


 92%|███████████████████████████████████████████████████████████████████████████████████▋       | 2201/2395 [38:04<03:21,  1.04s/it]

Got Errors {'c71f686d944f26b5a9b04ff37d69abff446fe1a2': 'Key c71f686d944f26b5a9b04ff37d69abff446fe1a2 not found in /da5_fast/All.sha1c/commit_71.tch'}


 92%|███████████████████████████████████████████████████████████████████████████████████▋       | 2202/2395 [38:05<03:20,  1.04s/it]

Got Errors {'c7660ea2f6d25907632240e3d58d5d425d0982e6': 'Key c7660ea2f6d25907632240e3d58d5d425d0982e6 not found in /da5_fast/All.sha1c/commit_71.tch'}


 92%|███████████████████████████████████████████████████████████████████████████████████▉       | 2209/2395 [38:12<03:13,  1.04s/it]

Got Errors {'c91355e717dec2c096ee03c9d02f4e999b76fe97': 'Key c91355e717dec2c096ee03c9d02f4e999b76fe97 not found in /da5_fast/All.sha1c/commit_73.tch'}


 93%|████████████████████████████████████████████████████████████████████████████████████▏      | 2217/2395 [38:20<03:04,  1.04s/it]

Got Errors {'cc3b50fa92fb0f610b78486c79cb1ce6ba688ce7': 'Key cc3b50fa92fb0f610b78486c79cb1ce6ba688ce7 not found in /da5_fast/All.sha1c/commit_76.tch'}


 93%|████████████████████████████████████████████████████████████████████████████████████▍      | 2223/2395 [38:27<02:58,  1.04s/it]

Got Errors {'cdc93e52fde506b783c8f116f1a1583e8aeb445d': 'Key cdc93e52fde506b783c8f116f1a1583e8aeb445d not found in /da5_fast/All.sha1c/commit_77.tch'}


 93%|████████████████████████████████████████████████████████████████████████████████████▊      | 2233/2395 [38:37<02:48,  1.04s/it]

Got Errors {'d11f7e42c3b417d80efb4e1acde5bc8cd33afa37': 'Key d11f7e42c3b417d80efb4e1acde5bc8cd33afa37 not found in /da5_fast/All.sha1c/commit_81.tch'}


 93%|█████████████████████████████████████████████████████████████████████████████████████      | 2238/2395 [38:42<02:43,  1.04s/it]

Got Errors {'d2548aca5c0a7ef620cf99f7f6f724e875e34ca5': 'Key d2548aca5c0a7ef620cf99f7f6f724e875e34ca5 not found in /da5_fast/All.sha1c/commit_82.tch'}


 94%|█████████████████████████████████████████████████████████████████████████████████████▎     | 2245/2395 [38:49<02:36,  1.04s/it]

Got Errors {'d47d00b0f7dc9b04a60484ac3fdb9ab295cb73ae': 'Key d47d00b0f7dc9b04a60484ac3fdb9ab295cb73ae not found in /da5_fast/All.sha1c/commit_84.tch'}


 94%|█████████████████████████████████████████████████████████████████████████████████████▌     | 2251/2395 [38:56<02:29,  1.04s/it]

Got Errors {'d625784d39d819b64c3ba3431a7b94e4cc1ebf76': 'Key d625784d39d819b64c3ba3431a7b94e4cc1ebf76 not found in /da5_fast/All.sha1c/commit_86.tch'}


 94%|█████████████████████████████████████████████████████████████████████████████████████▌     | 2252/2395 [38:57<02:28,  1.04s/it]

Got Errors {'d679b66b59f2da2019566b56e61fe00dafb81056': 'Key d679b66b59f2da2019566b56e61fe00dafb81056 not found in /da5_fast/All.sha1c/commit_86.tch'}


 95%|██████████████████████████████████████████████████████████████████████████████████████▏    | 2269/2395 [39:14<02:10,  1.04s/it]

Got Errors {'db9de92e4e72e9dceac8e5cf6f92cb0a6d659eac': 'Key db9de92e4e72e9dceac8e5cf6f92cb0a6d659eac not found in /da5_fast/All.sha1c/commit_91.tch'}


 95%|██████████████████████████████████████████████████████████████████████████████████████▎    | 2273/2395 [39:19<02:06,  1.04s/it]

Got Errors {'dc8838b8cd6dd60a266eb63886ddf7434007f509': 'Key dc8838b8cd6dd60a266eb63886ddf7434007f509 not found in /da5_fast/All.sha1c/commit_92.tch'}


 95%|██████████████████████████████████████████████████████████████████████████████████████▌    | 2279/2395 [39:25<02:00,  1.04s/it]

Got Errors {'debdd0e8206e4b66904ed7da36c49154f41aabea': 'Key debdd0e8206e4b66904ed7da36c49154f41aabea not found in /da5_fast/All.sha1c/commit_94.tch'}


 96%|███████████████████████████████████████████████████████████████████████████████████████    | 2291/2395 [39:37<01:47,  1.04s/it]

Got Errors {'e1bcab2b118f41e2e2fc72ca672d93721bf30d51': 'Key e1bcab2b118f41e2e2fc72ca672d93721bf30d51 not found in /da5_fast/All.sha1c/commit_97.tch'}


 96%|███████████████████████████████████████████████████████████████████████████████████████▌   | 2304/2395 [39:51<01:34,  1.04s/it]

Got Errors {'e58f84c782b8f2dc6370455912e010874a9e45ce': 'Key e58f84c782b8f2dc6370455912e010874a9e45ce not found in /da5_fast/All.sha1c/commit_101.tch'}


 97%|███████████████████████████████████████████████████████████████████████████████████████▉   | 2313/2395 [40:00<01:25,  1.04s/it]

Got Errors {'e834383e1894d3aa91eaa6ef83e680ba77b59d7e': 'Key e834383e1894d3aa91eaa6ef83e680ba77b59d7e not found in /da5_fast/All.sha1c/commit_104.tch'}


 97%|███████████████████████████████████████████████████████████████████████████████████████▉   | 2314/2395 [40:01<01:24,  1.04s/it]

Got Errors {'e86c25f3b5886b28a1d36d29f5f31932853dc435': 'Key e86c25f3b5886b28a1d36d29f5f31932853dc435 not found in /da5_fast/All.sha1c/commit_104.tch'}


 97%|████████████████████████████████████████████████████████████████████████████████████████▏  | 2320/2395 [40:07<01:18,  1.04s/it]

Got Errors {'ea17a25fde6b4707ac3cbbd55ece203fff6f8b7a': 'Key ea17a25fde6b4707ac3cbbd55ece203fff6f8b7a not found in /da5_fast/All.sha1c/commit_106.tch', 'ea2315194616245776d02b50062b9b74bcc1c49c': 'Key ea2315194616245776d02b50062b9b74bcc1c49c not found in /da5_fast/All.sha1c/commit_106.tch'}


 97%|████████████████████████████████████████████████████████████████████████████████████████▎  | 2323/2395 [40:11<01:14,  1.04s/it]

Got Errors {'eb13fe5b463a6753983624421fa5f17066ee6c26': 'Key eb13fe5b463a6753983624421fa5f17066ee6c26 not found in /da5_fast/All.sha1c/commit_107.tch'}


 98%|████████████████████████████████████████████████████████████████████████████████████████▊  | 2337/2395 [40:25<01:00,  1.04s/it]

Got Errors {'eed398b52786ca44ce1dd6df07e3143409ac3586': 'Key eed398b52786ca44ce1dd6df07e3143409ac3586 not found in /da5_fast/All.sha1c/commit_110.tch'}


 98%|████████████████████████████████████████████████████████████████████████████████████████▉  | 2341/2395 [40:29<00:56,  1.04s/it]

Got Errors {'f02f85b5f08870dd93b4b01e5253e8ec7cb6f869': 'Key f02f85b5f08870dd93b4b01e5253e8ec7cb6f869 not found in /da5_fast/All.sha1c/commit_112.tch'}


 98%|█████████████████████████████████████████████████████████████████████████████████████████▎ | 2351/2395 [40:40<00:45,  1.04s/it]

Got Errors {'f2f0cf04e9ad369c7284eaf8b540de633b2405d6': 'Key f2f0cf04e9ad369c7284eaf8b540de633b2405d6 not found in /da5_fast/All.sha1c/commit_114.tch'}


 98%|█████████████████████████████████████████████████████████████████████████████████████████▎ | 2352/2395 [40:41<00:44,  1.04s/it]

Got Errors {'f32695fef99d3154f5d0e3ef0ab5172d944f9114': 'Key f32695fef99d3154f5d0e3ef0ab5172d944f9114 not found in /da5_fast/All.sha1c/commit_115.tch', 'f34451f7a5aa7f743f9fe1b109d69791a3aba97a': 'Key f34451f7a5aa7f743f9fe1b109d69791a3aba97a not found in /da5_fast/All.sha1c/commit_115.tch'}


 98%|█████████████████████████████████████████████████████████████████████████████████████████▌ | 2357/2395 [40:46<00:39,  1.04s/it]

Got Errors {'f49b205dd85d7cb875659cd05ea74627a09ae56c': 'Key f49b205dd85d7cb875659cd05ea74627a09ae56c not found in /da5_fast/All.sha1c/commit_116.tch'}


 99%|█████████████████████████████████████████████████████████████████████████████████████████▊ | 2364/2395 [40:53<00:32,  1.04s/it]

Got Errors {'f6c14c6476c15facf2502d1bd392d4b2b4a24a01': 'Key f6c14c6476c15facf2502d1bd392d4b2b4a24a01 not found in /da5_fast/All.sha1c/commit_118.tch'}


100%|██████████████████████████████████████████████████████████████████████████████████████████▉| 2394/2395 [41:24<00:01,  1.04s/it]

Got Errors {'ffa7b9e11516e37286f5b7fca34bb2e9d05f2b8e': 'Key ffa7b9e11516e37286f5b7fca34bb2e9d05f2b8e not found in /da5_fast/All.sha1c/commit_127.tch'}


100%|███████████████████████████████████████████████████████████████████████████████████████████| 2395/2395 [41:25<00:00,  1.04s/it]


,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,004a56b528d6d1de376454a8cb0fa6701dc9ea71,a5b29a89ea7ff0f7d1029c6284542647e998c5b7,[c0fc90ad8c8f892779bb4e85dca216eb41b58bac],Chiron <Chiron@localhost>,1441289801,+0200,Chiron <Chiron@localhost>,1441289801,+0200,from future print compatible with conservation...,004a56b528d6d1de376454a8cb0fa6701dc9ea71,delsuc_spike
1,00749be973f4145c4b5c085c1a5dc08ba30c74d4,abd98dd01b91d2e09e416fc9f09510bec35a500a,[6b98df47dd1e5a50032bcdb704d17dc0b6fa8552],Chiron <Chiron@localhost>,1441358966,+0200,Chiron <Chiron@localhost>,1441358966,+0200,correction on erroneous name with self\n,00749be973f4145c4b5c085c1a5dc08ba30c74d4,delsuc_spike


In [6]:
# what does the first record looks like?
# commit  parents
v = commit_data[0]
# commit tree, [parent{s}], [author, unixtime, tz ], [ committer, unixtime, tz ], Commit_message ]
print(v)

{'commit': '004a56b528d6d1de376454a8cb0fa6701dc9ea71', 'tree': 'a5b29a89ea7ff0f7d1029c6284542647e998c5b7', 'parent': ['c0fc90ad8c8f892779bb4e85dca216eb41b58bac'], 'author': 'Chiron <Chiron@localhost>', 'author_time': 1441289801, 'author_tz': '+0200', 'committer': 'Chiron <Chiron@localhost>', 'committer_time': 1441289801, 'committer_tz': '+0200', 'message': 'from future print compatible with conservation of the doc\n'}


In [7]:
# Now that the data is retrieved, save it and commit to your GH fork
# First, we need to flatten info in order to export as csv
# our dataframe will have columns project,'commit, author, time, message'
dfinf = pd.DataFrame(columns=['project', 'commit', 'author', 'time', 'message'])
for k in commit_data:
  row = pd.Series({'project':prj, 'commit': k['commit'], 'author': k['author'], 'time': k['author_time'], 'message':k['message']})
  dfinf = pd.concat([dfinf, row.to_frame().T ], ignore_index=True)


In [8]:
# check if it has the right content
dfinf.head(1)

,project,commit,author,time,message
0,lsst_daf_butler,004a56b528d6d1de376454a8cb0fa6701dc9ea71,Chiron <Chiron@localhost>,1441289801,from future print compatible with conservation...


In [9]:
#mode a means append, so you have all your projects in the same file
yournetid='ksaravan'
dfinf.to_csv(yournetid+'_project_summary.csv', index=False,sep=';', mode='a', header=False)

In [4]:
import pandas as pd
df_commit_data = pd.read_csv('df_commit_data.csv')
df_commit_data.head()

,commit,tree,parent,author,author_time,author_tz,committer,committer_time,committer_tz,message,sha1,project
0,004a56b528d6d1de376454a8cb0fa6701dc9ea71,a5b29a89ea7ff0f7d1029c6284542647e998c5b7,['c0fc90ad8c8f892779bb4e85dca216eb41b58bac'],Chiron <Chiron@localhost>,1441289801,200,Chiron <Chiron@localhost>,1441289801,200,from future print compatible with conservation...,004a56b528d6d1de376454a8cb0fa6701dc9ea71,delsuc_spike
1,00749be973f4145c4b5c085c1a5dc08ba30c74d4,abd98dd01b91d2e09e416fc9f09510bec35a500a,['6b98df47dd1e5a50032bcdb704d17dc0b6fa8552'],Chiron <Chiron@localhost>,1441358966,200,Chiron <Chiron@localhost>,1441358966,200,correction on erroneous name with self\n,00749be973f4145c4b5c085c1a5dc08ba30c74d4,delsuc_spike
2,007fd7f2b3b9db0fa22d335a3070e2485d490d46,a59f2a8d1818cd4c2d7ab9b3bb9c6359a108a97c,['9b78b3c64bdcf4daa15c13e4d8d8a7c27a5dc3cf'],MAD <madelsuc@unistra.fr>,1620122959,200,MAD <madelsuc@unistra.fr>,1620122959,200,"improvement on documentation, and NMR Notebooks\n",007fd7f2b3b9db0fa22d335a3070e2485d490d46,delsuc_spike
3,00ca02e9ff99183b75793796c8c45b9df922f95d,dd6cfdc938eaf378171958b9fc612204a311fa26,['a543f8b58ca9869eeed5525c9f1fdecf73424ed4'],Marc-Andre' Delsuc <madelsuc@unistra.fr>,1444555855,200,Marc-Andre' Delsuc <madelsuc@unistra.fr>,1444555855,200,seems to work fine !\n,00ca02e9ff99183b75793796c8c45b9df922f95d,delsuc_spike
4,00f7f82c1a6cc30e789cd45edabcabc37723ed33,2aa2577654f785c418563ab4bccc8c54b23ccd6d,['13d0fd0a2966fb5a147fe12acfdb1dd53350734f'],MAD <madelsuc@unistra.fr>,1639069505,100,MAD <madelsuc@unistra.fr>,1639069505,100,improved spectral superimposing\n,00f7f82c1a6cc30e789cd45edabcabc37723ed33,delsuc_spike


In [31]:
# Updating the project summary correctly 
dfinf = df_commit_data[
    ['project', 'commit', 'author', 'author_time', 'message']
].copy()

dfinf = dfinf.rename(columns={'project': 'project_wocid', 'commit': 'commit_sha1', 'author_time': 'time', 'message': 'commit message'})

# Overwrite the incorrect summary file
dfinf.to_csv(
    'ksaravan_project_summary.csv',
    index=False,
    sep=';', 
    header=False
)

dfinf.head()

,project_wocid,commit_sha1,author,time,commit message
0,delsuc_spike,004a56b528d6d1de376454a8cb0fa6701dc9ea71,Chiron <Chiron@localhost>,1441289801,from future print compatible with conservation...
1,delsuc_spike,00749be973f4145c4b5c085c1a5dc08ba30c74d4,Chiron <Chiron@localhost>,1441358966,correction on erroneous name with self\n
2,delsuc_spike,007fd7f2b3b9db0fa22d335a3070e2485d490d46,MAD <madelsuc@unistra.fr>,1620122959,"improvement on documentation, and NMR Notebooks\n"
3,delsuc_spike,00ca02e9ff99183b75793796c8c45b9df922f95d,Marc-Andre' Delsuc <madelsuc@unistra.fr>,1444555855,seems to work fine !\n
4,delsuc_spike,00f7f82c1a6cc30e789cd45edabcabc37723ed33,MAD <madelsuc@unistra.fr>,1639069505,improved spectral superimposing\n


In [32]:
dfinf['project_wocid'].value_counts()

project_wocid
lsst_daf_butler             8810
mothur_mothur               4610
gyyang_neurogym             3031
papenfusslab_gridss         2382
delsuc_spike                1109
dswah_pygam                 1080
stephenslab_mashr            943
jbytecode_linregoutliers     902
arm61_pylj                   550
nasa_ccdd                    428
Name: count, dtype: int64

# Make sure you check in to your fork not just the notebook but also the csv files!!!

# Don't forget to add requested data from github and this notebook
### For each of the 10 projects go to their github repo and get the number of stars, number of forks, and the last commit date
### Report (in your notebook) the number of commits, the number of authors, and max and min time for each project based on WoC commits and also add the info you obtained from github

## 4. Manual GitHub Data Collection

|Project |Stars |Forks |Last Commit Date |
|---|---:|---:|---|
| delsuc/spike |21 |7 |August 10, 2026 |
| gyyang/neurogym |343 |59 |March 31, 2026 |
| PapenfussLab/gridss |285 |73 |May 21, 2025 |
| jbytecode/LinRegOutliers |53 |6 |September 4, 2026 |
| nasa/CCDD |93 |33 |June 30, 2026 |
| arm61/pylj |31 |22 |September 26, 2026 |
| mothur/mothur |281 |111 |August 21, 2026 |
| stephenslab/mashr |101 |21 |November 7, 2024 |
| dswah/pyGAM |1017 |292 |January 22, 2026 |
| lsst/daf_butler |16 |20 |September 27, 2026 |

## 5th Task

In [33]:
print(type(df_commit_data))
print(df_commit_data.columns.tolist())

<class 'pandas.DataFrame'>
['commit', 'tree', 'parent', 'author', 'author_time', 'author_tz', 'committer', 'committer_time', 'committer_tz', 'message', 'sha1', 'project']


In [34]:
# Using WoC API for getting the information

df_commit_data = pd.read_csv('df_commit_data.csv')

statistics = df_commit_data.groupby('project').agg({'commit': 'nunique', 'author': 'nunique', 'author_time': ['min', 'max']})

statistics

commit  author author_time            
                         nunique nunique         min         max
project                                                         
arm61_pylj                   550      17  1517503095  1733104946
delsuc_spike                1109      33  1416251831  1762186964
dswah_pygam                 1080      80  1484798062  1760626177
gyyang_neurogym             3031      39  1550764500  1761577669
jbytecode_linregoutliers     902      10  1598126055  1761947018
lsst_daf_butler             8810      86  1412152689  1762283255
mothur_mothur               4610      50  1232041010  1768327073
nasa_ccdd                    428      13  1484077520  1734639707
papenfusslab_gridss         2382      43  1377588244  1755694478
stephenslab_mashr            943      15  1481984097  1730994282

## Final Table

|Project |Commits |Authors |Min time|Max Time |Stars |Forks |Last Commit Date |
|---|---:|---:|---:|---:|---:|---:|---|
| delsuc/spike |1109 |33 |1416251831 |1762186964 |21 |7 |August 10, 2026 |
| gyyang/neurogym |3031	|39 |1550764500 |1761577669 |343 |59 |March 31, 2026 |
| PapenfussLab/gridss |2382	|43 |1377588244 |1755694478 |285 |73 |May 21, 2025 |
| jbytecode/LinRegOutliers |902 |10 |1598126055 |1761947018 |53 |6 |September 4, 2026 |
| nasa/CCDD |428 |13 |1484077520 |1734639707 |93 |33 |June 30, 2026 |
| arm61/pylj |550 |17 |1517503095 |1733104946 |31 |22 |September 26, 2026 |
| mothur/mothur |4610 |50 |1232041010 |1768327073 |281 |111 |August 21, 2026 |
| stephenslab/mashr |943 |15 |1481984097 |1730994282 |101 |21 |November 7, 2024 |
| dswah/pyGAM |1080 |80 |1484798062 |1760626177 |1017 |292 |January 22, 2026 |
| lsst/daf_butler |8810	|86 |1412152689 |1762283255 |16 |20 |September 27, 2026 |


## Part 2


In [2]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv(
    'ksaravan_project_summary.csv',
    sep=';',
    header=None,
    names=['project_wocid', 'commit_sha1', 'author', 'time', 'commit message']
)

df.head()

,project_wocid,commit_sha1,author,time,commit message
0,delsuc_spike,004a56b528d6d1de376454a8cb0fa6701dc9ea71,Chiron <Chiron@localhost>,1441289801,from future print compatible with conservation...
1,delsuc_spike,00749be973f4145c4b5c085c1a5dc08ba30c74d4,Chiron <Chiron@localhost>,1441358966,correction on erroneous name with self\n
2,delsuc_spike,007fd7f2b3b9db0fa22d335a3070e2485d490d46,MAD <madelsuc@unistra.fr>,1620122959,"improvement on documentation, and NMR Notebooks\n"
3,delsuc_spike,00ca02e9ff99183b75793796c8c45b9df922f95d,Marc-Andre' Delsuc <madelsuc@unistra.fr>,1444555855,seems to work fine !\n
4,delsuc_spike,00f7f82c1a6cc30e789cd45edabcabc37723ed33,MAD <madelsuc@unistra.fr>,1639069505,improved spectral superimposing\n
